lib/translator/inquiry.cc

```diff
@@ -0,0 +1,254 @@
+      result.page_length = kEui64Len;
+
+      // convert 64-bit number into hex string
+      sprintf(hex_string, "%08lx", identify_ns.eui64);
+    }
+
+    // insert _ and . in the correct positions
+    for (int i = 4; i < result.page_length - 1; i += 5) {
+      product_serial_number[i] = '_';
+    }
+    product_serial_number[result.page_length - 1] = '.';
+
+    // insert numbers
+    int pos = 0;
+    for (int i = 0; i < result.page_length - 1; ++i) {
+      if (product_serial_number[i] != '_') {
+        product_serial_number[i] = hex_string[pos++];
+      }
+    }
+  } else {
+    // 6.1.3.1.3
+    // valid for NVMe 1.0 devices only
+
+    result.page_length = kV1SerialLen;
+
+    // PRODUCT SERIAL NUMBER should be formed as follows:
+    // Bits 239:80 20 bytes of Serial Number (bytes 23:04 of Identify
+    // Controller data structure)
+    static_assert(sizeof(identify_ctrl.sn) == 20);
+    static_assert(sizeof(product_serial_number) >= 20);
+    memcpy(&product_serial_number, identify_ctrl.sn, sizeof(identify_ctrl.sn));
+
+    // Bits 79:72 ASCII representation of “_"
+    product_serial_number[sizeof(identify_ctrl.sn)] = '_';
+
+    // Bits 71:08 ASCII representation of 32 bit Namespace Identifier (NSID)
+    sprintf(&product_serial_number[sizeof(identify_ctrl.sn) + 1], "%08x", nsid);
+
+    // Bits 07:00 ASCII representation of “."
+    product_serial_number[kV1SerialLen - 1] = '.';
+  }
+
+  WriteValue(result, buffer);
+  WriteValue(product_serial_number, buffer.subspan(sizeof(result)));
+}
+}  // namespace
+
+StatusCode InquiryToNvme(absl::Span<const uint8_t> raw_scsi,
+                         nvme_defs::GenericQueueEntryCmd& identify_ns,
+                         nvme_defs::GenericQueueEntryCmd& identify_ctrl,
+                         uint32_t& alloc_len, scsi_defs::LunAddress lun) {
+  scsi_defs::InquiryCommand cmd = {};
+  if (!ReadValue(raw_scsi, cmd)) return StatusCode::kFailure;
+
+  alloc_len = cmd.allocation_length;
+
+  // Allocate prp & assign to command
+  uint64_t ns_prp = AllocPages(1);
+  if (!ns_prp) return StatusCode::kFailure;
+
+  identify_ns = nvme_defs::GenericQueueEntryCmd{
+      .opc = static_cast<uint8_t>(nvme_defs::AdminOpcode::kIdentify),
+      .nsid = lun};
+  identify_ns.dptr.prp.prp1 = ns_prp;
+  identify_ns.cdw[0] = 0x0;  // cns val
+
+  uint64_t ctrl_prp = AllocPages(1);
+  if (!ctrl_prp) return StatusCode::kFailure;
+
+  identify_ctrl = nvme_defs::GenericQueueEntryCmd{
+      .opc = static_cast<uint8_t>(nvme_defs::AdminOpcode::kIdentify),
+  };
+  identify_ctrl.dptr.prp.prp1 = ctrl_prp;
+  identify_ctrl.cdw[0] = 0x1;  // cns val
+  return StatusCode::kSuccess;
+}
+
+// Main logic engine for the Inquiry command
+StatusCode InquiryToScsi(
+    absl::Span<const uint8_t> raw_scsi, absl::Span<uint8_t> buffer,
+    absl::Span<const nvme_defs::GenericQueueEntryCmd> nvme_cmds) {
+  scsi_defs::InquiryCommand inquiry_cmd = {};
+
+  if (!ReadValue(raw_scsi, inquiry_cmd)) return StatusCode::kFailure;
```

Comment:
Nit: might be more appropriate to return `StatusCode::kInvalidInput` . In these casting checks, we've also been doing `DebugLog("Malformed Inquiry Command")`